daisy/step_create_images.go

```diff
@@ -16,63 +16,124 @@ package daisy
+func (ci *CreateImages) populate(ctx context.Context, s *Step) dErr {
 	var errs dErr
-	for _, i := range *c {
-		errs = addErrs(errs, i.populate(ctx, s))
+	if ci.Images != nil {
+		for _, i := range ci.Images {
+			errs = addErrs(errs, populate(ctx, i, &i.ImageBase, s))
+		}
 	}
+
+	if ci.ImagesBeta != nil {
+		for _, i := range ci.ImagesBeta {
+			errs = addErrs(errs, populate(ctx, i, &i.ImageBase, s))
+		}
+	}
+
 	return errs
 }
 
-func (c *CreateImages) validate(ctx context.Context, s *Step) dErr {
+func (ci *CreateImages) validate(ctx context.Context, s *Step) dErr {
 	var errs dErr
-	for _, i := range *c {
-		errs = addErrs(errs, i.validate(ctx, s))
+
+	if usesBetaFeatures(ci.ImagesBeta) {
+		if ci.ImagesBeta != nil {
+			for _, i := range ci.ImagesBeta {
+				errs = addErrs(errs, validate(ctx, i, &i.ImageBase, i.Licenses, s))
+			}
+		}
+	} else {
+		if ci.Images != nil {
+			for _, i := range ci.Images {
+				errs = addErrs(errs, validate(ctx, i, &i.ImageBase, i.Licenses, s))
+			}
+		}
 	}
+
 	return errs
 }
 
-func (c *CreateImages) run(ctx context.Context, s *Step) dErr {
+func (ci *CreateImages) run(ctx context.Context, s *Step) dErr {
 	var wg sync.WaitGroup
 	w := s.w
 	e := make(chan dErr)
-	for _, i := range *c {
-		wg.Add(1)
-		go func(ci *Image) {
-			defer wg.Done()
-			// Get source disk link if SourceDisk is a daisy reference to a disk.
-			if d, ok := w.disks.get(ci.SourceDisk); ok {
-				ci.SourceDisk = d.link
-			}
 
-			// Delete existing if OverWrite is true.
-			if ci.OverWrite {
-				// Just try to delete it, a 404 here indicates the image doesn't exist.
-				if err := w.ComputeClient.DeleteImage(ci.Project, ci.Name); err != nil {
-					if apiErr, ok := err.(*googleapi.Error); !ok || apiErr.Code != 404 {
-						e <- errf("error deleting existing image: %v", err)
-						return
-					}
+	createImage := func(ci ImageInterface, overwrite bool) {
+		defer wg.Done()
+		// Get source disk link if SourceDisk is a daisy reference to a disk.
+		if d, ok := w.disks.get(ci.getSourceDisk()); ok {
+			ci.setSourceDisk(d.link)
+		}
+
+		// Delete existing if OverWrite is true.
+		if overwrite {
+			// Just try to delete it, a 404 here indicates the image doesn't exist.
+			if err := ci.delete(w.ComputeClient); err != nil {
+				if apiErr, ok := err.(*googleapi.Error); !ok || apiErr.Code != 404 {
+					e <- errf("error deleting existing image: %v", err)
+					return
 				}
 			}
+		}
 
-			w.LogStepInfo(s.name, "CreateImages", "Creating image %q.", ci.Name)
-			if err := w.ComputeClient.CreateImage(ci.Project, &ci.Image); err != nil {
-				e <- newErr(err)
-				return
-			}
-		}(i)
+		w.LogStepInfo(s.name, "CreateImages", "Creating image %q.", ci.getName())
+		if err := ci.create(w.ComputeClient); err != nil {
+			e <- newErr(err)
+			return
+		}
+	}
+
+	if usesBetaFeatures(ci.ImagesBeta) {
+		for _, i := range ci.ImagesBeta {
+			wg.Add(1)
+			go createImage(i, i.OverWrite)
```

Comment:
Passing in overwrite is a bit redundant as you already are passing in ImageInterface

Reply:
I didn't want to add another method to the interface since overwrite is part of ImageBase. That's why I pass ImageBase reference in addition to ImageInterface to certain methods, so I can use ImageBase fields without having to add getters/setters in Image Interface.